test/Dialect/Torch/simplify-shape-calculations.mlir

```diff
@@ -489,3 +493,36 @@ func.func @shape_calc_with_two_uses(%arg0: !torch.vtensor<[2],f32>) -> !torch.vt
 
   return %arg0 : !torch.vtensor<[2],f32>
 }
+
+// CHECK-LABEL: func.func @propagate_sizes_through_copy(
+// CHECK-SAME:      %[[ARG:.*]]: !torch.vtensor<[2],f32>) -> !torch.tensor {
+// CHECK:         %[[TWO:.*]] = torch.constant.int 2
+// CHECK:         %[[CAST_0:.*]] = torch.tensor_static_info_cast %[[ARG]] : !torch.vtensor<[2],f32> to !torch.vtensor
+// CHECK:         %[[COPY:.*]] = torch.copy.to_tensor %[[CAST_0]] : !torch.tensor
+// CHECK:         %[[SHAPE_CALC:.*]] = torch.shape.calculate {
+// CHECK:           %[[CAST_1:.*]] = torch.tensor_static_info_cast %[[COPY]] : !torch.tensor to !torch.vtensor<[2],f32>
+// CHECK:           %[[NEG:.*]] = torch.aten.neg %[[CAST_1]] : !torch.vtensor<[2],f32> -> !torch.tensor<[2],unk>
+// CHECK:           torch.shape.calculate.yield %[[NEG]] : !torch.tensor<[2],unk>
+// CHECK:         } shapes {
+// CHECK:           %[[SHAPE_LIST:.*]] = torch.prim.ListConstruct %[[TWO]] : (!torch.int) -> !torch.list<int>
+// CHECK:           torch.shape.calculate.yield.shapes %[[SHAPE_LIST]] : !torch.list<int>
+// CHECK:         } : !torch.tensor<[2],unk>
+// CHECK:         %[[CAST_2:.*]] = torch.tensor_static_info_cast %[[SHAPE_CALC]] : !torch.tensor<[2],unk> to !torch.tensor
+// CHECK:         return %[[CAST_2]] : !torch.tensor
+// CHECK:       }
+
+func.func @propagate_sizes_through_copy(%arg: !torch.vtensor<[2],f32>) -> !torch.tensor {
+  %cast = torch.tensor_static_info_cast %arg : !torch.vtensor<[2],f32> to !torch.vtensor
+  %tensor = torch.copy.to_tensor %cast : !torch.tensor
+
+  %shape_calc = torch.shape.calculate {
+    %neg = torch.aten.neg %tensor : !torch.tensor -> !torch.tensor
```

Comment:
Generally we shouldn't be doing shape inference without fully converting the program to value semantics. I think that would make torch.copy.to_tensor go away, and I think that the tensor_static_info_cast should canonicalize away in that case.

Reply:
I see, thanks for the tip!  Perhaps I should be focusing on that pass.  There's a `torch.overwrite.tensor.contents` in the IR which is preventing the copies from being removed.  I'll look there first and come back to this.

Reply:
I've noticed that we've had a few issues surrounding this imperfect conversion to value semantics causing deceptive error messages for developers.

Could you maybe add a pass lib/Dialect/Torch/Transforms/VerifyConversionToValueSemantics.cpp that checks that all tensors have been converted to value semantics, and put it in the pipeline after MaximizeValueSemantics (code can be mostly copypasted from lib/Dialect/TorchConversion/Transforms/VerifyInvariantsBeforeBackendLowering.cpp). I think there might be an edge case around return ops that we need to handle specially, but otherwise I think copying that other pass should be all we need.

Reply:
I found that all tensor-returning functions end in the following sequence:

```
%copy = torch.copy.to_tensor %result : !torch.tensor
return %copy : !torch.tensor
```

which causes the invariant to fail.  Would you suggest updating the `GlobalizeObjectGraph` to generate functions that return value tensors instead?

Reply:
No, I would special-case the verification of that construct, or run RefinePublicReturn just before (we will also need to run it again later in the pipeline where it is today though).

Reply:
Perfect, thanks for educating me (I had no idea what `RefinePublicReturn` did until you mentioned it)!  The validation pass is in PR https://github.com/llvm/torch-mlir/pull/1052.